File: apps/desktop-client/src-tauri/src/windows/window_manager.rs
/// 统一窗口管理器
/// 负责协调所有窗口的创建、配置和生命周期管理
/// 重点解决焦点管理问题：确保焦点始终在 main-focus 窗口上

use log::{info, error, warn};
use tauri::{AppHandle, Manager};

use crate::windows::{
    main_focus_window::MainFocusWindow,
    control_bar_window::ControlBarWindow,
    close_button_window::CloseButtonWindow,
    main_content_window::MainContentWindow,
};

pub struct WindowManager {
    app_handle: AppHandle,
    main_focus_window: MainFocusWindow,
    control_bar_window: ControlBarWindow,
    close_button_window: CloseButtonWindow,
    main_content_window: MainContentWindow,
}

impl WindowManager {
    pub fn new(app_handle: AppHandle) -> Self {
        let main_focus_window = MainFocusWindow::new(app_handle.clone());
        let control_bar_window = ControlBarWindow::new(app_handle.clone());
        let close_button_window = CloseButtonWindow::new(app_handle.clone());
        let main_content_window = MainContentWindow::new(app_handle.clone());

        Self {
            app_handle,
            main_focus_window,
            control_bar_window,
            close_button_window,
            main_content_window,
        }
    }

    /// 初始化所有窗口和 NSPanel 配置
    /// 关键：创建 main-focus 主焦点窗口并确保焦点管理正确
    pub async fn initialize(&mut self) -> Result<(), String> {
        info!("开始初始化窗口管理器");

        // 1. 首先创建 main-focus 主焦点窗口，这是最重要的
        match self.main_focus_window.create().await {
            Ok(_) => {
                info!("main-focus 主焦点窗口已创建");
            }
            Err(e) => {
                error!("创建 main-focus 窗口失败，这是致命错误: {}", e);
                return Err(format!("创建 main-focus 窗口失败: {}", e));
            }
        }

        info!("准备配置 control-bar 为 NSPanel...");
        
        // 2. 配置 control-bar 为 NSPanel（非阻塞，失败不影响启动）
        match self.control_bar_window.setup_as_panel() {
            Ok(_) => {
                info!("control-bar NSPanel 配置成功");
            }
            Err(e) => {
                warn!("配置 control-bar NSPanel 失败，但不影响启动: {}", e);
            }
        }

        info!("control-bar 配置完成，准备配置 close-button...");

        // 3. 配置 close-button 为 NSPanel（非阻塞，失败不影响启动）
        // 由于 control-bar 已经设置了 center: true，close-button 会自动定位
        match self.close_button_window.setup_as_panel() {
            Ok(_) => {
                info!("close-button NSPanel 配置成功");
            }
            Err(e) => {
                warn!("配置 close-button NSPanel 失败，但不影响启动: {}", e);
            }
        }

        info!("close-button 配置完成，准备配置 main-content...");

        // 4. 配置 main-content 为 NSPanel（非阻塞，失败不影响启动）
        match self.main_content_window.setup_as_panel() {
            Ok(_) => {
                info!("main-content NSPanel 配置成功");
            }
            Err(e) => {
                warn!("配置 main-content NSPanel 失败，但不影响启动: {}", e);
            }
        }

        info!("main-content 配置完成，准备恢复焦点...");

        // 5. 最后确保焦点在 main-focus 窗口上（非阻塞）
        if let Err(e) = self.ensure_main_focus() {
            warn!("初始化时恢复 main-focus 焦点失败，但不影响启动: {}", e);
        }

        info!("窗口管理器初始化完成");
        Ok(())
    }

    /// 核心方法：确保焦点始终在 main-focus 窗口上
    /// 这是解决焦点切换问题的关键
    pub fn ensure_main_focus(&self) -> Result<(), String> {
        // 立即恢复焦点到 main-focus 窗口
        self.main_focus_window.ensure_focus()?;
        
        // 记录日志以便调试
        info!("🔥 强制恢复焦点到 main-focus 主焦点窗口");
        
        Ok(())
    }

    /// 创建主内容窗口（现在只是确保存在，因为窗口已在 tauri.conf.json 中定义）
    pub async fn create_main_content(&mut self) -> Result<(), String> {
        // 确保 main-focus 主焦点窗口存在
        if !self.main_focus_window.is_created() {
            self.main_focus_window.create().await?;
        }

        // 主内容窗口已在 tauri.conf.json 中定义，无需手动创建
        info!("主内容窗口已在配置中定义，无需手动创建");
        
        // 关键：确保焦点在 main-focus 窗口上
        self.ensure_main_focus()?;
        
        Ok(())
    }

    /// 显示主内容窗口
    pub fn show_main_content(&self) -> Result<(), String> {
        self.main_content_window.show()?;
        
        // 关键：主内容窗口显示后立即恢复焦点到 main-focus
        self.ensure_main_focus()?;
        
        Ok(())
    }

    /// 隐藏主内容窗口
    pub fn hide_main_content(&self) -> Result<(), String> {
        self.main_content_window.hide()?;
        
        // 确保焦点还是在 main-focus
        self.ensure_main_focus()?;
        
        Ok(())
    }

    /// 切换主内容窗口
    pub fn toggle_main_content(&self) -> Result<(), String> {
        self.main_content_window.toggle()?;
        
        // 关键：无论显示还是隐藏，都要确保焦点在 main-focus
        self.ensure_main_focus()?;
        
        Ok(())
    }

    /// 显示所有浮动窗口
    pub fn show_floating_windows(&self) -> Result<(), String> {
        let mut success_count = 0;
        let mut error_count = 0;

        // 显示 control-bar
        if let Err(e) = self.control_bar_window.show() {
            error!("显示 control-bar 失败: {}", e);
            error_count += 1;
        } else {
            success_count += 1;
        }

        // 显示 close-button
        if let Err(e) = self.close_button_window.show() {
            error!("显示 close-button 失败: {}", e);
            error_count += 1;
        } else {
            success_count += 1;
        }

        // 关键：显示浮动窗口后确保焦点在 main-focus
        self.ensure_main_focus()?;

        if error_count == 0 {
            info!("所有浮动窗口显示成功，成功: {}", success_count);
            Ok(())
        } else {
            warn!("部分浮动窗口显示失败，成功: {}, 失败: {}", success_count, error_count);
            Ok(()) // 不返回错误，允许部分成功
        }
    }

    /// 隐藏所有浮动窗口
    pub fn hide_floating_windows(&self) -> Result<(), String> {
        let mut success_count = 0;
        let mut error_count = 0;

        // 隐藏 control-bar
        if let Err(e) = self.control_bar_window.hide() {
            error!("隐藏 control-bar 失败: {}", e);
            error_count += 1;
        } else {
            success_count += 1;
        }

        // 隐藏 close-button
        if let Err(e) = self.close_button_window.hide() {
            error!("隐藏 close-button 失败: {}", e);
            error_count += 1;
        } else {
            success_count += 1;
        }

        // 即使隐藏了，焦点还是要在 main-focus
        self.ensure_main_focus()?;

        if error_count == 0 {
            info!("所有浮动窗口隐藏成功，成功: {}", success_count);
            Ok(())
        } else {
            warn!("部分浮动窗口隐藏失败，成功: {}, 失败: {}", success_count, error_count);
            Ok(()) // 不返回错误，允许部分成功
        }
    }

    /// 切换浮动窗口显示状态
    pub fn toggle_floating_windows(&self) -> Result<(), String> {
        // 检查 control-bar 的可见性作为参考
        if let Some(window) = self.app_handle.get_webview_window("control-bar") {
            match window.is_visible() {
                Ok(is_visible) => {
                    if is_visible {
                        self.hide_floating_windows()
                    } else {
                        self.show_floating_windows()
                    }
                }
                Err(e) => {
                    error!("获取浮动窗口状态失败: {}", e);
                    Err(format!("获取浮动窗口状态失败: {}", e))
                }
            }
        } else {
            Err("control-bar 窗口不存在".to_string())
        }
    }

}

File: apps/desktop-client/src/main/windows/WindowManager.ts
import type { BrowserWindow, Tray } from 'electron';
import { screen } from 'electron';
import type { AppState } from '../../shared/types.js';
import { createLogger } from '../../utils/logger.js';

const log = createLogger('WindowManager');
import { ScreenshotWatcher } from '../utils/screenshotWatcher.js';
import { WebSocketClient } from '../websocket/WebSocketClient.js';
import { AIQuestionHistoryWindow } from './AIQuestionHistoryWindow.js';
import { AIQuestionWindow } from './AIQuestionWindow.js';
import { ControlBarWindow } from './ControlBarWindow.js';
import { InterviewerWindow } from './InterviewerWindow.js';
import { MainContentWindow } from './MainContentWindow.js';
import { TrayMenuWindow } from './TrayMenuWindow.js';

export class WindowManager {
  private controlBarWindow: ControlBarWindow;
  private mainContentWindow!: MainContentWindow; // 在 initialize 中初始化
  private aiQuestionWindow!: AIQuestionWindow; // 在 initialize 中初始化
  private webSocketClient: WebSocketClient;
  private isDevelopment: boolean;
  private appState: AppState;
  private aiQuestionHistoryWindow!: AIQuestionHistoryWindow; // 在 initialize 中初始化
  private interviewerWindow!: InterviewerWindow; // 在 initialize 中初始化
  private trayMenuWindow!: TrayMenuWindow; // 托盘菜单窗口
  private windowStatesBeforeHide: {
    isMainContentVisible: boolean;
    isAIQuestionVisible: boolean;
    isAIQuestionHistoryVisible: boolean;
    isInterviewerVisible: boolean;
  } = {
    isMainContentVisible: false,
    isAIQuestionVisible: false,
    isAIQuestionHistoryVisible: false,
    isInterviewerVisible: false,
  };
  private hasEverBeenHidden: boolean = false;

  private aiWindowHeightPercentage: number = 75; // 默认 75%
  private screenshotWatcher: ScreenshotWatcher | null = null;

  constructor(isDevelopment: boolean = false) {
    this.isDevelopment = isDevelopment;

    // 初始化应用状态
    this.appState = {
      isControlBarVisible: true,
      isCloseButtonVisible: false, // 关闭按钮状态由组件内部管理
      isMainContentVisible: false,
      isAIQuestionVisible: false,
      isAIQuestionHistoryVisible: false as any,
    };

    // 创建窗口实例 - control-bar 现在作为主焦点窗口，关闭按钮已集成
    this.controlBarWindow = new ControlBarWindow(this.isDevelopment);

    // 创建 WebSocket 客户端
    this.webSocketClient = new WebSocketClient(this);
  }

  /**
   * 初始化窗口管理器
   */
  public async initialize(): Promise<void> {
    try {
      // 1. 创建主内容窗口（作为主窗口）
      this.mainContentWindow = new MainContentWindow(this.isDevelopment, null);
      await this.mainContentWindow.create();

      // 2. 创建控制条窗口
      await this.controlBarWindow.create();

      // 3. 创建子窗口，以 control-bar 作为父窗口
      const controlBarBrowserWindow = this.controlBarWindow.getBrowserWindow();
      this.aiQuestionWindow = new AIQuestionWindow(this.isDevelopment, controlBarBrowserWindow);
      this.aiQuestionHistoryWindow = new AIQuestionHistoryWindow(
        this.isDevelopment,
        controlBarBrowserWindow,
      );
      this.interviewerWindow = new InterviewerWindow(this.isDevelopment, controlBarBrowserWindow);

      // 4. 创建 AI 问答窗口（初始隐藏）
      await this.aiQuestionWindow.create();

      // 5. 创建 AI 问答历史窗口（初始隐藏）
      await this.aiQuestionHistoryWindow.create();

      // 6. 创建 Interviewer 窗口（初始隐藏）
      await this.interviewerWindow.create();

      // 6.5 创建托盘菜单窗口（初始隐藏，不设置 parent 避免影响 ControlBar 窗口状态）
      this.trayMenuWindow = new TrayMenuWindow(this.isDevelopment, null);
      await this.trayMenuWindow.create();

      // 7. 设置控制条窗口关闭回调，隐藏所有子窗口
      this.controlBarWindow.setOnCloseCallback(() => {
        this.hideAllChildWindows();
      });

      // 8. 设置窗口事件监听
      this.setupWindowEvents();

      // 9. 连接 WebSocket 客户端
      this.webSocketClient.connect();

      // 10. 显示浮动窗口（control-bar）
      this.showFloatingWindows();

      // 11. 监听屏幕分辨率变化
      this.setupScreenEvents();

      // 12. 启动截图/录屏监控：检测到立即隐藏子窗口；结束后恢复
      this.screenshotWatcher = new ScreenshotWatcher();
      this.screenshotWatcher.onChange((capturing) => {
        try {
          if (capturing) {
            // 统一走已有的保存+隐藏逻辑：包含 control-bar、main-content 以及所有子窗口
            this.hideFloatingWindows();
          } else {
            // 截图结束，按保存的状态一次性恢复
            this.showFloatingWindows();
          }
        } catch {}
      });
      this.screenshotWatcher.start();
    } catch (error) {
      log.error('initialize', '窗口管理器初始化失败', {}, error);
      throw error;
    }
  }

  /**
   * 设置屏幕事件监听
   */
  private setupScreenEvents(): void {
    // 监听显示器变化事件
    screen.on('display-added', () => {
      this.updateAIWindowSize();
    });

    screen.on('display-removed', () => {
      this.updateAIWindowSize();
    });

    screen.on('display-metrics-changed', () => {
      this.updateAIWindowSize();
    });
  }

  /**
   * 更新 AI 窗口尺寸
   */
  private updateAIWindowSize(): void {
    if (this.appState.isAIQuestionVisible) {
      this.aiQuestionWindow.updateWindowSize();
    }
  }

  /**
   * 设置窗口事件监听
   */
  private setupWindowEvents(): void {
    // 监听控制条窗口事件
    const controlBarWindow = this.controlBarWindow.getBrowserWindow();
    if (controlBarWindow) {
      controlBarWindow.on('focus', () => {});
    }

    // 监听主内容窗口事件
    const mainContentWindow = this.mainContentWindow.getBrowserWindow();
    if (mainContentWindow) {
      // 允许 main-content 在交互期间获得焦点（以便键盘输入）。
      mainContentWindow.on('focus', () => {});

      mainContentWindow.on('blur', () => {
        setTimeout(() => this.ensureMainFocus(), 0);
      });

      mainContentWindow.on('hide', () => {
        setTimeout(() => this.ensureMainFocus(), 0);
      });

      mainContentWindow.on('close', (event) => {
        // 阻止窗口关闭，改为隐藏
        event.preventDefault();
        this.hideMainContent();
      });
    }

    // 监听 AI 问答窗口事件用于联动历史窗口和 interviewer 窗口定位
    const aiWindow = this.aiQuestionWindow.getBrowserWindow();
    if (aiWindow) {
      const relayoutSideWindows = () => {
        if (this.isAIQuestionHistoryVisible()) {
          const historyBounds = this.computeHistoryBoundsNextToAI();
          if (historyBounds) {
            this.aiQuestionHistoryWindow.setBounds(historyBounds);
          }
        }
        if (this.isInterviewerVisible()) {
          const interviewerBounds = this.computeInterviewerBoundsNextToAI();
          if (interviewerBounds) {
            this.interviewerWindow.setBounds(interviewerBounds);
          }
        }
      };

      aiWindow.on('move', relayoutSideWindows);
      aiWindow.on('moved', relayoutSideWindows as any);
      aiWindow.on('resize', relayoutSideWindows);
      aiWindow.on('resized', relayoutSideWindows as any);
      aiWindow.on('show', relayoutSideWindows);
    }
  }

  /**
   * 确保焦点在主焦点窗口上（现在是 control-bar）
   */
  public ensureMainFocus(): void {
    this.controlBarWindow.ensureFocus();
  }

  /**
   * 显示浮动窗口（恢复隐藏前的状态）
   */
  public showFloatingWindows(): void {
    // 显示控制条窗口
    this.controlBarWindow.show();
    this.appState.isControlBarVisible = true;
    this.appState.isCloseButtonVisible = true;

    // 广播应用可见性变化
    const { broadcastAppVisibilityChanged } = require('../ipc/handlers.js');
    broadcastAppVisibilityChanged(true);

    // 通知菜单更新
    try {
      const { ipcMain } = require('electron');
      ipcMain.emit('update-tray-menu');
    } catch {}

    // 如果从未隐藏过，恢复到当前实际状态；否则恢复到隐藏前的状态
    if (!this.hasEverBeenHidden) {
      // 第一次显示，恢复当前实际状态
      if (this.appState.isMainContentVisible) {
        this.mainContentWindow.show();
      }
      if (this.appState.isAIQuestionVisible) {
        this.aiQuestionWindow.show();
      }
      if (this.isAIQuestionHistoryVisible()) {
        this.showAIQuestionHistoryNextToAI();
      }
      if (this.isInterviewerVisible()) {
        this.showInterviewerNextToAI();
      }
    } else {
      // 恢复隐藏前的状态
      if (this.windowStatesBeforeHide.isMainContentVisible) {
        this.mainContentWindow.show();
        this.appState.isMainContentVisible = true;
      }
      if (this.windowStatesBeforeHide.isAIQuestionVisible) {
        this.aiQuestionWindow.show();
        this.appState.isAIQuestionVisible = true;
      }
      if (this.windowStatesBeforeHide.isAIQuestionHistoryVisible) {
        this.showAIQuestionHistoryNextToAI();
      }
      if (this.windowStatesBeforeHide.isInterviewerVisible) {
        this.showInterviewerNextToAI();
      }
    }

    // 初始时确保焦点在 control-bar
    setTimeout(() => this.ensureMainFocus(), 100);
  }

  /**
   * 隐藏浮动窗口（保存当前状态）
   */
  public hideFloatingWindows(): void {
    // 保存所有窗口当前的实际可见状态
    this.windowStatesBeforeHide = {
      isMainContentVisible: this.mainContentWindow.isVisible(),
      isAIQuestionVisible: this.aiQuestionWindow.isVisible(),
      isAIQuestionHistoryVisible: this.isAIQuestionHistoryVisible(),
      isInterviewerVisible: this.isInterviewerVisible(),
    };
    this.hasEverBeenHidden = true;

    // 隐藏所有子窗口
    this.hideAllChildWindows();

    // 隐藏 main-content 窗口
    if (this.mainContentWindow.isVisible()) {
      this.mainContentWindow.hide();
      this.appState.isMainContentVisible = false;
    }

    // 隐藏控制条窗口
    this.controlBarWindow.hide();
    this.appState.isControlBarVisible = false;
    this.appState.isCloseButtonVisible = false;

    // 广播应用可见性变化
    const { broadcastAppVisibilityChanged } = require('../ipc/handlers.js');
    broadcastAppVisibilityChanged(false);

    // 通知菜单更新
    try {
      const { ipcMain } = require('electron');
      ipcMain.emit('update-tray-menu');
    } catch {}
  }

  /**
   * 切换浮动窗口显示状态
   */
  public toggleFloatingWindows(): void {
    if (this.appState.isControlBarVisible) {
      this.hideFloatingWindows();
    } else {
      this.showFloatingWindows();
    }
  }

  /**
   * 显示关闭按钮 - 由控制条组件内部管理
   */
  public showCloseButton(): void {
    this.appState.isCloseButtonVisible = true;
  }

  /**
   * 隐藏关闭按钮 - 由控制条组件内部管理
   */
  public hideCloseButton(): void {
    this.appState.isCloseButtonVisible = false;
  }

  /**
   * 显示主内容窗口
   */
  public showMainContent(): void {
    this.mainContentWindow.show();
    this.appState.isMainContentVisible = true;
    // 主内容窗口已显示
  }

  /**
   * 隐藏主内容窗口
   */
  public hideMainContent(): void {
    this.mainContentWindow.hide();
    this.appState.isMainContentVisible = false;
    // 主内容窗口已隐藏

    // 立即恢复焦点
    this.ensureMainFocus();
  }

  /**
   * 切换主内容窗口显示状态
   */
  public toggleMainContent(): void {
    if (this.appState.isMainContentVisible) {
      this.hideMainContent();
    } else {
      this.showMainContent();
    }
  }

  /**
   * 打开外部链接
   */
  public openExternalUrl(url: string): void {
    const { shell } = require('electron');
    shell.openExternal(url);
  }

  /**
   * 获取应用状态
   */
  public getAppState(): AppState {
    return { ...this.appState };
  }

  /**
   * 获取控制条窗口实例
   */
  public getControlBarWindow(): BrowserWindow | null {
    return this.controlBarWindow.getBrowserWindow();
  }

  /**
   * 获取 AI 问答窗口实例
   */
  public getAIQuestionWindow(): AIQuestionWindow {
    return this.aiQuestionWindow;
  }

  /**
   * 获取主内容窗口实例
   */
  public getMainContentWindow(): MainContentWindow {
    return this.mainContentWindow;
  }

  /**
   * 显示 AI 问答窗口
   */
  public showAIQuestion(): void {
    this.aiQuestionWindow.show();
    this.appState.isAIQuestionVisible = true;
  }

  /**
   * 隐藏 AI 问答窗口
   */
  public hideAIQuestion(): void {
    this.aiQuestionWindow.hide();
    this.appState.isAIQuestionVisible = false;
    // 关闭 AI 窗口时，同步关闭历史窗口
    if (this.isAIQuestionHistoryVisible()) {
      this.aiQuestionHistoryWindow.hide();
    }
  }

  /**
   * 切换 AI 问答窗口显示状态
   */
  public toggleAIQuestion(): void {
    if (this.appState.isAIQuestionVisible) {
      this.hideAIQuestion();
    } else {
      this.showAIQuestion();
    }
  }

  // === 历史窗口相关 ===
  public isAIQuestionHistoryVisible(): boolean {
    const w = this.aiQuestionHistoryWindow.getBrowserWindow();
    return !!(w && w.isVisible());
  }

  private computeHistoryBoundsNextToAI(): Electron.Rectangle | null {
    const aiWin = this.aiQuestionWindow.getBrowserWindow();
    if (!aiWin) return null;
    const aiBounds = aiWin.getBounds();
    const display = screen.getDisplayMatching(aiBounds) || screen.getPrimaryDisplay();
    const workArea = display.workArea;
    const rightSpace = workArea.x + workArea.width - (aiBounds.x + aiBounds.width);
    const minWidth = 260;
    const maxWidth = Math.max(420, Math.floor(workArea.width * 0.5));
    const spacing = Math.floor(rightSpace * 0.05); // 5% 的 rightSpace 作为间距
    let width = Math.floor(rightSpace * 0.9);
    width = Math.max(minWidth, Math.min(width, maxWidth, rightSpace - spacing));
    const x = aiBounds.x + aiBounds.width + spacing;
    const y = aiBounds.y;
    const height = aiBounds.height;
    return { x, y, width, height };
  }

  public showAIQuestionHistoryNextToAI(): void {
    const bounds = this.computeHistoryBoundsNextToAI();
    if (!bounds) return;
    this.aiQuestionHistoryWindow.setBounds(bounds);
    this.aiQuestionHistoryWindow.show();
  }

  public hideAIQuestionHistory(): void {
    this.aiQuestionHistoryWindow.hide();
  }

  public toggleAIQuestionHistoryNextToAI(): void {
    if (this.isAIQuestionHistoryVisible()) {
      this.hideAIQuestionHistory();
    } else {
      this.showAIQuestionHistoryNextToAI();
    }
  }

  // === Interviewer 窗口相关 ===
  public isInterviewerVisible(): boolean {
    const w = this.interviewerWindow.getBrowserWindow();
    return !!(w && w.isVisible());
  }

  private computeInterviewerBoundsNextToAI(): Electron.Rectangle | null {
    const aiWin = this.aiQuestionWindow.getBrowserWindow();
    if (!aiWin) return null;
    const aiBounds = aiWin.getBounds();
    const display = screen.getDisplayMatching(aiBounds) || screen.getPrimaryDisplay();
    const workArea = display.workArea;
    const leftSpace = aiBounds.x - workArea.x;
    const minWidth = 260;
    const maxWidth = Math.max(420, Math.floor(workArea.width * 0.5));
    const spacing = Math.floor(leftSpace * 0.05); // 5% 的 leftSpace 作为间距
    let width = Math.floor(leftSpace * 0.9);
    width = Math.max(minWidth, Math.min(width, maxWidth, leftSpace - spacing));
    const x = Math.max(workArea.x, aiBounds.x - spacing - width);
    const y = aiBounds.y;
    const height = aiBounds.height;
    return { x, y, width, height };
  }

  public showInterviewerNextToAI(): void {
    const bounds = this.computeInterviewerBoundsNextToAI();
    if (!bounds) return;
    this.interviewerWindow.setBounds(bounds);
    this.interviewerWindow.show();
  }

  public hideInterviewer(): void {
    this.interviewerWindow.hide();
  }

  public toggleInterviewerNextToAI(): void {
    if (this.isInterviewerVisible()) {
      this.hideInterviewer();
    } else {
      this.showInterviewerNextToAI();
    }
  }

  /**
   * 设置 AI 窗口高度百分比
   */
  public setAIWindowHeightPercentage(percentage: number): void {
    this.aiWindowHeightPercentage = percentage;
    this.aiQuestionWindow.setHeightPercentage(percentage);

    // 同步更新其他窗口
    if (this.isAIQuestionHistoryVisible()) {
      this.showAIQuestionHistoryNextToAI();
    }
    if (this.isInterviewerVisible()) {
      this.showInterviewerNextToAI();
    }
  }

  /**
   * 获取 AI 窗口高度百分比
   */
  public getAIWindowHeightPercentage(): number {
    return this.aiWindowHeightPercentage;
  }

  /**
   * 切换 AI 窗口模式
   */
  public switchToMode(
    mode: 'voice-qa' | 'mock-interview' | 'interview-training',
    jobId?: string,
  ): void {
    // 向 AI Question 窗口发送模式切换事件
    if (this.aiQuestionWindow.getBrowserWindow()) {
      this.aiQuestionWindow.getBrowserWindow()!.webContents.send('mode-change', mode, jobId);
    }

    // 向 AI Question History 窗口发送模式切换事件
    if (this.aiQuestionHistoryWindow.getBrowserWindow()) {
      this.aiQuestionHistoryWindow.getBrowserWindow()!.webContents.send('mode-change', mode, jobId);
    }

    // 向 Interviewer 窗口发送模式切换事件
    if (this.interviewerWindow.getBrowserWindow()) {
      this.interviewerWindow.getBrowserWindow()!.webContents.send('mode-change', mode, jobId);
    }
  }

  /**
   * 设置"提问 AI"按钮的禁用状态
   */
  public setAskAIButtonDisabled(disabled: boolean): void {
    // 向 control-bar 窗口发送按钮禁用状态事件
    if (this.controlBarWindow.getBrowserWindow()) {
      this.controlBarWindow
        .getBrowserWindow()!
        .webContents.send('ask-ai-button-disabled', disabled);
    }
  }

  /**
   * 获取 WebSocket 客户端实例
   */
  public getWebSocketClient(): WebSocketClient {
    return this.webSocketClient;
  }

  /**
   * 隐藏所有子窗口（保持状态）
   */
  public hideAllChildWindows(): void {
    if (this.appState.isAIQuestionVisible) {
      this.aiQuestionWindow.hide();
      this.appState.isAIQuestionVisible = false;
    }

    if (this.isAIQuestionHistoryVisible()) {
      this.aiQuestionHistoryWindow.hide();
    }

    if (this.isInterviewerVisible()) {
      this.interviewerWindow.hide();
    }
  }

  /**
   * 在隐藏状态下准备打开面试窗口
   * 不直接显示窗口，而是更新状态，等待恢复时显示
   */
  public prepareInterviewWindowsWhileHidden(): void {
    // 更新隐藏前状态，标记这些窗口应该在恢复时显示
    this.windowStatesBeforeHide.isAIQuestionVisible = true;
    this.windowStatesBeforeHide.isAIQuestionHistoryVisible = true;
    this.windowStatesBeforeHide.isInterviewerVisible = true;
    this.hasEverBeenHidden = true;
  }

  /**
   * 销毁窗口管理器
   */

  /**
   * 切换点击穿透模式 - 供快捷键调用
   */
  public toggleClickThroughMode(): void {
    try {
      const toggleFn = (global as any).toggleClickThroughMode;
      if (typeof toggleFn === 'function') {
        toggleFn();
      } else {
        log.error('toggleClickThroughMode', 'toggleClickThroughMode 函数未找到');
      }
    } catch (error) {
      log.error('toggleClickThroughMode', '调用点击穿透切换函数失败', {}, error);
    }
  }

  public destroy(): void {
    // 停止截图监听器
    if (this.screenshotWatcher) {
      this.screenshotWatcher.stop();
    }

    // 断开 WebSocket 连接
    this.webSocketClient.disconnect();

    this.controlBarWindow.destroy();
    this.mainContentWindow.destroy();
    this.aiQuestionWindow.destroy();
    this.aiQuestionHistoryWindow.destroy();
    this.interviewerWindow.destroy();
    this.trayMenuWindow.destroy();
  }

  /**
   * 显示托盘菜单窗口
   */
  public showTrayMenu(tray: Tray | null): void {
    if (!this.trayMenuWindow) return;
    this.trayMenuWindow.showNearTray(tray);
  }

  /**
   * 隐藏托盘菜单窗口
   */
  public hideTrayMenu(): void {
    if (!this.trayMenuWindow) return;
    this.trayMenuWindow.hide();
  }

  /**
   * 切换托盘菜单窗口显示状态
   */
  public toggleTrayMenu(tray: Tray | null): void {
    if (!this.trayMenuWindow) return;
    if (this.trayMenuWindow.isVisible()) {
      this.trayMenuWindow.hide();
    } else {
      this.trayMenuWindow.showNearTray(tray);
    }
  }

  /**
   * 通知托盘菜单设置已变更
   */
  public notifyTrayMenuSettingsChanged(): void {
    if (this.trayMenuWindow) {
      this.trayMenuWindow.notifySettingsChanged();
    }
  }
}
